### PR TITLE
Adds gnosis integration for existing hh task/script

### PR DESCRIPTION
**Pull request by ShubhamChndrvnshi**

### CODE REVIEW

File: gnosis/helper.js [hunk 1]
```diff
@@ -0,0 +1,13 @@
+
+async function submitTransactionToGnosisSafe(DEPLOYER_SAFE, artifact, functionName, ...args) {
```
Comment (KohliSuraj) [inline gnosis/helper.js]: This is excellent.

This is exactly what we needed `const tx = await artifact.connect(signer)[functionName](...args)` 👍

File: gnosis/helper.js [hunk 2]
```diff
@@ -0,0 +1,14 @@
...
```
Comment (KohliSuraj) [inline gnosis/helper.js]: We can rename DEPLOYER_SAFE to follow camel case to signify that it's a parameter.

File: gnosis/signer.js [hunk 3]
```diff
@@ -0,0 +1,18 @@
+const Safe = require('@safe-global/safe-core-sdk')
+const { SafeEthersSigner, SafeService } = require('@safe-global/safe-ethers-adapters')
+const ethAdapter = require('./adapter')
+
+async function getGnosisSigner(DEPLOYER_SAFE){
```
Comment (KohliSuraj) [inline gnosis/signer.js]: We can rename DEPLOYER_SAFE to follow camel case to signify that it's a parameter.

File: gnosis/helper.js [hunk 4]
```diff
@@ -0,0 +1,14 @@
...
+    if (!ethers.utils.isAddress(DEPLOYER_SAFE)) throw new Error("Invalid safe address")
+    const getGnosisSigner = require('./signer');
+    const colors = require('colors');
+    const signer = await getGnosisSigner(DEPLOYER_SAFE)
+    hre.network.name == 'dashboard' ? console.log(colors.blue("OPEN TRUFFLE DASHBOARD TO SIGN TXN")) : console.log(colors.green("Signing Tx with gnosis signer"))
+    const tx = await artifact.connect(signer)[functionName](...args)
+    console.log("Gnosis tx hash: ", tx.hash)
+    console.log(`Go to gnosis dashbaord to view/confirm the txn: https://app.safe.global/transactions/queue?safe=${DEPLOYER_SAFE}`)
```
Comment (KohliSuraj) [inline gnosis/helper.js]: minor type: dashbaord

File: tasks/add-impl-to-finder.js [hunk 5]
```diff
@@ -4,8 +4,16 @@ task("add-impl-to-finder", "Adds assets to Asset Manager")
...
     .addFlag("skipIfImplementationExists", "boolean value to skip changing implementation if it already exists")
+    .addOptionalParam(
+        "gnosisSafe",
+        "Gnosis safe address, should be given if trasactions need to be submitted to gnosis",
+        undefined,
+        types.string
+    )
     .setAction(
         async (args, hre) => {
+            if (args.gnosisSafe && !ethers.utils.isAddress(args.gnosisSafe)) throw new Error("Invalid safe address")
```
Comment (KohliSuraj) [inline tasks/add-impl-to-finder.js]: we can do this check on line 36 itself. No need to do it once here and then again on line 36..

File: tasks/clay-approve.js [hunk 6]
```diff
@@ -2,22 +2,32 @@
...
             const { expect } = require('chai');
             const { deployer } = await hre.getNamedAccounts();
             const { getTxUrl } = require('../utils/helper');
+            const submitTransactionToGnosisSafe = require('../gnosis/helper');
 
             const clayToken = await ethers.getContract("ClayToken", deployer);
             console.log("Using ClayToken: ", clayToken.address);
 
-            const clayBalance = await clayToken.balanceOf(deployer);
+            const clayBalance = await clayToken.balanceOf(args.gnosisSafe || deployer);
```
Comment (KohliSuraj) [inline tasks/clay-approve.js]: TODO: need to improve this. Did you hear anything back from safe about using gnosis signer to call read only functions?

File: tasks/close-asset.js [hunk 7]
```diff
@@ -33,6 +42,7 @@ task("close-asset", "Closes assets on asset manager")
                 console.log("Closing swap pair...")
 
                 try {
+                    if (gnosisSafe) return submitTransactionToGnosisSafe(gnosisSafe, assetManager, 'closeSwapPair', args.id);
```
Comment (KohliSuraj) [inline tasks/close-asset.js]: shouldn't this be `args.gnosisSafe`?

File: tasks/close-asset.js [hunk 8]
```diff
@@ -47,6 +57,7 @@ task("close-asset", "Closes assets on asset manager")
                 console.log("Closing staking reward...")
 
                 try {
+                    if (gnosisSafe) return submitTransactionToGnosisSafe(gnosisSafe, assetManager, 'closeStakingReward', args.id);
```
Comment (KohliSuraj) [inline tasks/close-asset.js]: shouldn't this be `args.gnosisSafe`?

**Review by KohliSuraj — CHANGES_REQUESTED** [review]
@ShubhamChndrvnshi Code looks good, have added some feedback.

As discussed earlier, in order to test and complete this, we should transfer ownership of contracts deployed as part of user testing v0.1.2 and then perform operations/tasks on them via this gnosis integration.

File: tasks/close-asset.js (hunk 8 above)
Comment (ShubhamChndrvnshi) [reply tasks/close-asset.js]: yes

File: tasks/close-asset.js (hunk 7 above)
Comment (ShubhamChndrvnshi) [reply tasks/close-asset.js]: yes

File: tasks/clay-approve.js (hunk 6 above)
Comment (ShubhamChndrvnshi) [reply tasks/clay-approve.js]: nope

File: tasks/add-impl-to-finder.js (hunk 5 above)
Comment (ShubhamChndrvnshi) [reply tasks/add-impl-to-finder.js]: cleaned it

File: gnosis/helper.js (hunk 4 above)
Comment (ShubhamChndrvnshi) [reply gnosis/helper.js]: corrected

File: gnosis/signer.js (hunk 3 above)
Comment (ShubhamChndrvnshi) [reply gnosis/signer.js]: done in latest commit

File: gnosis/helper.js (hunk 2 above)
Comment (ShubhamChndrvnshi) [reply gnosis/helper.js]: done in latest commit

File: gnosis/signer.js [hunk 9]
```diff
@@ -2,21 +2,21 @@ const Safe = require('@safe-global/safe-core-sdk')
 const { SafeEthersSigner, SafeService } = require('@safe-global/safe-ethers-adapters')
 const ethAdapter = require('./adapter')
 
-const serviceURLObj = {
+const chaidIdToServiceUrl = {
```
Comment (KohliSuraj) [inline gnosis/signer.js]: minor typo: chaidIdToServiceUrl

File: gnosis/signer.js [hunk 10]
```diff
@@ -2,21 +2,21 @@ const Safe = require('@safe-global/safe-core-sdk')
...
         const chainId = await ethAdapter.getChainId()
         console.log("Initializing safe: ", deployerSafeAddress)
-        const service = new SafeService(serviceURLObj[chainId])
+        const service = new SafeService(chaidIdToServiceUrl[chainId])
         const safe = await Safe.default.create({ ethAdapter: ethAdapter, safeAddress: deployerSafeAddress })
         console.log("Safe created")
 
-        const safeSigner = new SafeEthersSigner(safe, service, hre.network.provider)
-        console.log("Safe singer initialized")
+        const safeSigner = new SafeEthersSigner(safe, service, new hre.ethers.providers.Web3Provider(hre.network.provider))
```
Comment (KohliSuraj) [inline gnosis/signer.js:18]: Awesome! This change is crux for solving read only function calls.

**Review by KohliSuraj — APPROVED** [review]

File: gnosis/signer.js (hunk 9 above)
Comment (ShubhamChndrvnshi) [reply gnosis/signer.js]: fixed

File: gnosis/signer.js (hunk 10 above)
Comment (ShubhamChndrvnshi) [reply gnosis/signer.js:18]: yeah

**Pull request merged**